### DISCUSSION TITLE
How to add <meta> data for robots and googlebot?

### DISCUSSION THREAD

**Discussion by tikita** [datenstrom/community#676] [Ask a question] [answered]
Hello, how can I include in one or more page such meta tags?

Example:

```
<meta name="robots" content="noindex, nofollow, noarchive, nosnippet">
<meta name="googlebot" content="noindex, nofollow, noarchive, nosnippet">
```

Thank you, Tikita

**Comment by GiovanniSalmeri** [comment]
If you want to add the same meta data to all the pages of your site, you can simply edit header.html and add whatever you wish. If you want to add different meta data to each page, this is not so straightforward, but I can write a quick and dirty extension which allows this. Let me know!

**Reply by tikita** [reply]
> If you want to add the same meta data to all the pages of your site, you can simply edit header.html and add whatever you wish.

Yes, exactly, for all sides this is the easiest way.

> If you want to add different meta data to each page, this is not so straightforward, but I can write a quick and dirty extension which allows this. Let me know!

Great!

**Comment by GiovanniSalmeri** [comment]
Here is the little extension:

```
<?php
// Metatags extension, https://github.com/GiovanniSalmeri/yellow-metatags

class YellowMetatags {
    const VERSION = "0.8.15";
    public $yellow;         // access to API
    
    // Handle initialisation
    public function onLoad($yellow) {
        $this->yellow = $yellow;
        $this->yellow->system->setDefault("metatags", "color-scheme, google, googlebot, referrer, robots, theme-color"); // https://developer.mozilla.org/en-US/docs/Web/HTML/Element/meta/name
    }
    
    // Handle page extra data
    public function onParsePageExtra($page, $name) {
        $output = null;
        if ($name=="header") {
            $metaNames = preg_split("/\s*,\s*/", $this->yellow->system->get("metatags"));
            foreach ($metaNames as $metaName) {
                if ($this->yellow->page->isExisting($metaName)) {
                    $output .= "<meta name=\"".htmlspecialchars($metaName)."\" content=\"".htmlspecialchars($this->yellow->page->get($metaName))."\" />\n";
                }
            }
        }
        return $output;
    }
}

```

Create a file with this content, name it `metatags.php` and put it in `system/extensions`. Now you can add for example `Robots: noindex, nofollow, noarchive, nosnippet`in the page settings. Hope it is useful!

**Reply by markseuffert** [reply]
I was reading your response this morning and was really curious about how you would do this in a few lines. I didn't expect it to be that short. Well done, very useful.  👍

**Reply by GiovanniSalmeri** [reply]
Thank you! Can I publish it as a separate extension? or is it better to integrate it into the extension Meta? or something else?

**Reply by markseuffert** [reply]
I don't have a strong opinion on search engine robots and personally would use the `robots.txt` to keep them at bay. Not sure where's the exact line between publishing a separate extension and writing down an example extension. Seeing your extension is a good starting point for [learning more about the API](https://datenstrom.se/yellow/help/api-for-developers) and how to do more powerful things as a web developer. And as an API example I find it very useful. You can't have enough examples. 😀

**Reply by tikita** [reply]
> Here is the little extension (...)

Thank you very much for your quick solution!

But I did not understand where I _can add for example Robots: noindex, nofollow, noarchive, nosnippet **in the page settings**._ 

Where can I find the _page settings_ or which file needs to be adjusted?

Thanks and greetings, Tikita

**Reply by GiovanniSalmeri** [reply]
The page settings are at the top of the page between ---. For example:

```
---
Author: Immanuel Kant
Title: Kritik der reinen Vernunft
Robots: noindex, nofollow
---

Die menschliche Vernunft hat das besondere Schicksal in
einer Gattung ihrer Erkenntnisse: daß sie durch Fragen
belästigt wird, die sie nicht abweisen kann; denn sie sind ihr
durch die Natur der Vernunft selbst aufgegeben, die sie aber
auch nicht beantworten kann, denn sie übersteigen alles 
Vermögen der menschlichen Vernunft.
```

**Reply by tikita** [reply]
All right (it works)! Once again, many, many thanks ...

PS: And then Kant ;-)

**Reply by GiovanniSalmeri** [reply]
Bitteschön! :-)

**Comment by GiovanniSalmeri** [comment, marked as answer]
Now public: https://github.com/GiovanniSalmeri/yellow-metatags

**Comment by nzyeehaa** [comment]
@tikita 
I dont wish to subtract from @GiovanniSalmeri great work, but I use a very simplistic approach which doesnt need an extension...

Each of my sites has a unique theme / layout, so if the site needs it, I include this in the **/system/layouts/header.html**
`<meta name="robots" content="<?php echo $this->yellow->page->getHtml("robots") ?>" />`

and include this on the page front-matter if it requires it,
`Robots: noindex, nofollow, noarchive, nosnippet`

And if I want to include something on a _regular_ basis, I change the **/content/shared/page-new-default.md** so every **new** page already has the tag, and I delete what I dont need or delete the whole line.
```
---
Title: Page
Robots: noindex, nofollow, noarchive, nosnippet
---
This is a new page.
```

If you require both tags, the header would look like this...
```
<meta name="robots" content="<?php echo $this->yellow->page->getHtml("robots") ?>" />
<meta name="googlebot" content="<?php echo $this->yellow->page->getHtml("robots") ?>" />
```

Giovannis plugin would be better if you switch themes often
